[WIP] Messenger support

mynameisbogdan opened the pull request:
| Q | A
| --- | ---
| Branch? | 2.0 <!--choose version number-->
| Bug fix? | no
| New feature? | yes
| BC breaks? | no
| Deprecations? | no
| Tests pass? | yes
| Fixed tickets | <!-- #-pre-fixed issue number(s), if any -->
| License | MIT
| Doc PR | <!--highly recommended for new features-->

Adding optional `symfony/messenger` support.

Needs documentation.

peter-gribanov commented:
Perhaps you can use the [EnqueueBundle](https://symfony.com/doc/2.x/bundles/LiipImagineBundle/resolve-cache-images-in-background.html) documentation as a basis to document this feature.

alex-cool-tech commented:
Hi, any updates maybe?

comxd commented:
I have a problem using flysystem and Google Cloud Storage:

```
"error" => "Handling "App\Message\LiipImageResolveCacheMessage" failed: The mime type of file path/to/my/image.jpg?itok=yeDn2W0I must be image/xxx or application/pdf, got application/octet-stream."
```

dbu reviewed (approved):
thanks a lot, this looks good!

can you please add documentation (duplicate and adjust what we have for enqueue-bundle)?

then we can merge this, and we will deprecate the enqueue integration.

dbu commented on Message/Handler/ResolveCacheHandler.php:
```diff
@@ -0,0 +1,47 @@
+
+        foreach ($filters as $filter) {
+            if ($message->isForce()) {
+                $this->filterService->bustCache($path, $filter);
```
likely needs to be adjusted to the change in https://github.com/liip/LiipImagineBundle/pull/1347 once that is merged.

mynameisbogdan replied:
Ok, I'll change it then.

mynameisbogdan commented:
@dbu done regarding the docs.

dbu commented on Resources/doc/resolve-cache-images-in-background.rst:
```diff
@@ -17,10 +14,98 @@ Though there are some disadvantages:
   The varnish may cache the page with those links to the resolve controller.
   A browser keeps sending requests to it though there is no need for it after the first call.
 
+Messenger
```
```suggestion
Symfony Messenger
```

dbu commented on Resources/doc/resolve-cache-images-in-background.rst:
```diff
@@ -17,10 +14,98 @@ Though there are some disadvantages:
+Step 1: Install
+~~~~~~~~~~~~~~~~
+
+First, we have to `install symfony/messenger`_. You have to basically use composer to install the Symfony component.
```
```suggestion
First, `install symfony/messenger`_ with composer:
```

dbu commented on Resources/doc/resolve-cache-images-in-background.rst:
```diff
@@ -17,10 +14,98 @@ Though there are some disadvantages:
+                'Liip\ImagineBundle\Message\ResolveCache': async
+
+Step 2: Configure LiipImagineBundle
+~~~~~~~~~~~~~~~~
```
the underlines should be the same length as the title

dbu commented on Resources/doc/resolve-cache-images-in-background.rst:
```diff
@@ -17,10 +14,98 @@ Though there are some disadvantages:
+
+You have to dispatch a message in order to process images in background.
+The message must contain the original image path (in terms of LiipImagineBundle).
+If you do not define filters the background process will resolve cache for all available filters.
```
```suggestion
If you do not define filters, the background process will resolve cache for all available filters.
```

dbu commented on Resources/doc/resolve-cache-images-in-background.rst:
```diff
@@ -17,10 +14,98 @@ Though there are some disadvantages:
+You have to dispatch a message in order to process images in background.
+The message must contain the original image path (in terms of LiipImagineBundle).
+If you do not define filters the background process will resolve cache for all available filters.
+If the cache already exist the background process does recreate it by default
```
```suggestion
If the cache already exists, the background process does recreate it by default
```

dbu commented on Resources/doc/resolve-cache-images-in-background.rst:
```diff
@@ -17,10 +14,98 @@ Though there are some disadvantages:
+The message must contain the original image path (in terms of LiipImagineBundle).
+If you do not define filters the background process will resolve cache for all available filters.
+If the cache already exist the background process does recreate it by default
+You can force cache to be recreated and in this case the cached image is removed and a new one replaces it.
```
is the above line incorrect then? by default, we do *not* recreate the image, you need to force to recreate?

mynameisbogdan replied:
Yes, seems incorrect. We do not recreate the image, you have set force to true for that.

dbu commented on Resources/doc/resolve-cache-images-in-background.rst:
```diff
@@ -17,10 +14,98 @@ Though there are some disadvantages:
   The varnish may cache the page with those links to the resolve controller.
   A browser keeps sending requests to it though there is no need for it after the first call.
 
```
i think we should explain a bit more how the solution works here:

```suggestion
To prepare the cached images in advance, the LiipImagineBundle allows you to use a message queue to have a worker warm up the cache asynchronously. Your application has to send messages about the images as it becomes aware of them (file upload, import processes, ...) and you need to run the worker for the message queue.
```

dbu commented on Message/Handler/ResolveCacheHandler.php:
```diff
@@ -0,0 +1,47 @@
+use Liip\ImagineBundle\Service\FilterService;
+use Symfony\Component\Messenger\Handler\MessageHandlerInterface;
+
+/** @experimental */
```
```suggestion
/** 
 * Listen to ResolveCache messages and prepare the cache for those images.
 *
 * @experimental
 */
```

dbu commented on Message/ResolveCache.php:
```diff
@@ -0,0 +1,59 @@
+
+namespace Liip\ImagineBundle\Message;
+
+/** @experimental */
```
```suggestion
/**
 * Message to warm up the cache for an image.
 * 
 * Your application needs to dispatch this message when it becomes aware of a new image.
 *
 * @experimental 
 */
```

mynameisbogdan commented:
Great @dbu! 

I implemented your suggestions with one little change. I renamed the message to `WarmupCache`, and the handler to `WarmupCacheHandler` to stick to Symfony's naming convention of `${messageClassName}Handler`. 

Also in the docs, I changed some `resolve` wording to `warmup`. And recommended a new transport named `liip_imagine`.

It's this okay with you?

dbu commented on Resources/doc/resolve-cache-images-in-background.rst:
```diff
@@ -16,11 +13,100 @@ Though there are some disadvantages:
+To prepare the cached images in advance, the LiipImagineBundle allows you to use a message queue to have a worker warm up the cache asynchronously. Your application has to send messages about the images as it becomes aware of them (file upload, import processes, ...) and you need to run the worker for the message queue.
+
+Symfony Messenger
+-------------------
```
```suggestion
-----------------
```

dbu commented on Resources/doc/resolve-cache-images-in-background.rst:
```diff
@@ -16,11 +13,100 @@ Though there are some disadvantages:
+    $ php bin/console messenger:consume liip_imagine --time-limit=3600 --memory-limit=256M -vv
+
+Step 4: Send warmup cache message
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
```suggestion
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
